### PR TITLE
Port io::basp::worker

### PR DESCRIPTION
**Pull request by jakobod**
This PR ports the `basp::worker` scaffold which was formerly in `caf::io` to `caf::net`.

### CODE REVIEW

File: libcaf_net/src/application.cpp [hunk 1]
```diff
@@ -44,7 +44,10 @@ namespace caf {
 namespace net {
 namespace basp {
 
-application::application(proxy_registry& proxies) : proxies_(proxies) {
+application::application(proxy_registry& proxies)
+  : proxies_(proxies),
+    queue_{std::unique_ptr<message_queue>{new message_queue}},
```
Comment (Neverlord) [inline libcaf_net/src/application.cpp]:
```suggestion
    queue_(new message_queue),
```

File: libcaf_net/src/application.cpp [hunk 2]
```diff
@@ -44,7 +44,10 @@ namespace caf {
...
+    hub_{std::unique_ptr<hub_type>{new hub_type}} {
```
Comment (Neverlord) [inline libcaf_net/src/application.cpp]:
```suggestion
    hub_(new hub_type) {
```

File: libcaf_net/caf/net/basp/worker.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  /// Only the ::worker_hub has access to the construtor.
+  worker(hub_type& hub, message_queue& queue, proxy_registry& proxies);
+
+  ~worker() override;
+
+  // -- management -------------------------------------------------------------
+
+  void launch(const node_id& last_hop, const basp::header& hdr,
+              const buffer_type& payload);
```
Comment (Neverlord) [inline libcaf_net/caf/net/basp/worker.hpp]:
```suggestion
              const span<byte>& payload);
```

The implementation copies the content into `payload_`. There's no reason to require an actual buffer here.

File: libcaf_net/src/application.cpp [hunk 4]
```diff
@@ -248,34 +251,41 @@ error application::handle_handshake(packet_writer&, header hdr,
...
-    return ec::invalid_payload;
-  // Try to fetch the receiver.
-  auto dst_hdl = system().registry().get(dst_id);
-  if (dst_hdl == nullptr) {
-    CAF_LOG_DEBUG("no actor found for given ID, drop message");
-    return caf::none;
+  auto worker = hub_->pop();
+  // TODO: This copy is nessecary because the worker interface expects a buffer
+  // type...
+  buffer_type buf(payload.begin(), payload.end());
```
Comment (Neverlord) [inline libcaf_net/src/application.cpp]: The worker should instead accept a span (see comment in the worker header).

File: libcaf_net/src/application.cpp [hunk 5]
```diff
@@ -248,34 +251,41 @@ error application::handle_handshake(packet_writer&, header hdr,
...
+    worker->launch(node_id{}, hdr, buf);
+  } else {
+    CAF_LOG_DEBUG(
+      "out of BASP workers, continue deserializing an actor_message");
+    // If no worker is available then we have no other choice than to take
+    // the performance hit and deserialize in this thread.
+    struct handler : remote_message_handler<handler> {
+      handler(message_queue* queue, proxy_registry* proxies,
+              actor_system* system, node_id last_hop, basp::header& hdr,
+              buffer_type& payload)
```
Comment (Neverlord) [inline libcaf_net/src/application.cpp]:
```suggestion
             byte_span payload)
```

File: libcaf_net/src/application.cpp [hunk 6]
```diff
@@ -248,34 +251,41 @@ error application::handle_handshake(packet_writer&, header hdr,
...
+          hdr_(hdr),
+          payload_(payload) {
+        msg_id_ = queue_->new_id();
+      }
+      message_queue* queue_;
+      proxy_registry* proxies_;
+      actor_system* system_;
+      node_id last_hop_;
+      basp::header& hdr_;
+      buffer_type& payload_;
```
Comment (Neverlord) [inline libcaf_net/src/application.cpp]:
```suggestion
      byte_span payload_;
```

File: libcaf_net/src/application.cpp [hunk 7]
```diff
@@ -248,34 +251,41 @@ error application::handle_handshake(packet_writer&, header hdr,
...
-  if (dst_hdl == nullptr) {
-    CAF_LOG_DEBUG("no actor found for given ID, drop message");
-    return caf::none;
+  auto worker = hub_->pop();
+  // TODO: This copy is nessecary because the worker interface expects a buffer
+  // type...
+  buffer_type buf(payload.begin(), payload.end());
+  if (worker != nullptr) {
+    CAF_LOG_DEBUG("launch BASP worker for deserializing an actor_message");
+    worker->launch(node_id{}, hdr, buf);
```
Comment (Neverlord) [inline libcaf_net/src/application.cpp]:
```suggestion
    worker->launch(node_id{}, hdr, payload);
```

File: libcaf_net/src/application.cpp [hunk 8]
```diff
@@ -248,34 +251,41 @@ error application::handle_handshake(packet_writer&, header hdr,
...
+      }
+      message_queue* queue_;
+      proxy_registry* proxies_;
+      actor_system* system_;
+      node_id last_hop_;
+      basp::header& hdr_;
+      buffer_type& payload_;
+      uint64_t msg_id_;
+    };
+    handler f{queue_.get(), &proxies_, system_, node_id{}, hdr, buf};
```
Comment (Neverlord) [inline libcaf_net/src/application.cpp]:
```suggestion
    handler f{queue_.get(), &proxies_, system_, node_id{}, hdr, payload};
```

File: libcaf_net/src/worker.cpp [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+worker::~worker() {
+  // nop
+}
+
+// -- management ---------------------------------------------------------------
+
+void worker::launch(const node_id& last_hop, const basp::header& hdr,
+                    const buffer_type& payload) {
```
Comment (Neverlord) [inline libcaf_net/src/worker.cpp]:
```suggestion
                    span<byte> payload) {
```

File: libcaf_net/test/net/basp/message_queue.cpp [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+#include "caf/behavior.hpp"
+
+using namespace caf;
+
+namespace {
+
+behavior testee_impl() {
+  return {[](ok_atom, int) {
+    // nop
+  }};
```
Comment (Neverlord) [inline libcaf_net/test/net/basp/message_queue.cpp]:
```suggestion
  },};
```

Aside from other reasons, `clang-format` will format message handlers better if you're using a trailing comma.

File: libcaf_net/test/net/basp/worker.cpp [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+#include "caf/serializer_impl.hpp"
+
+using namespace caf;
+
+namespace {
+
+behavior testee_impl() {
+  return {[](ok_atom) {
+    // nop
+  }};
```
Comment (Neverlord) [inline libcaf_net/test/net/basp/worker.cpp]:
```suggestion
  },};
```

File: libcaf_net/test/net/basp/worker.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#define CAF_SUITE basp.worker
```
Comment (Neverlord) [inline libcaf_net/test/net/basp/worker.cpp]:
```suggestion
#define CAF_SUITE net.basp.worker
```

File: libcaf_net/test/net/basp/message_queue.cpp [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#define CAF_SUITE basp.message_queue
```
Comment (Neverlord) [inline libcaf_net/test/net/basp/message_queue.cpp]:
```suggestion
#define CAF_SUITE net.basp.message_queue
```

**Comment by Neverlord** [comment]
@jakobod removing the extra copy, did you compare the performance to `caf_io` again? Is it still slower than `caf_io`?

**Comment by jakobod** [comment]
@Neverlord I haven't. But will do that today.

**Pull request merged**